osgbuild/fetch_sources.py

```diff
@@ -32,194 +82,263 @@
     logging.basicConfig(format="%(message)s", level=logging.INFO)
 
 
-def process_meta_url(line, destdir, nocheck):
-    """
-    Process a serialized URL spec.  Should be of the format:
-     type=git url=https://github.com/opensciencegrid/cvmfs-config-osg.git name=cvmfs-config-osg tag=0.1 hash=e2b54cd1b94c9e3eaee079490c9d85f193c52249
-    'name' can be derived from the URL if the last component in the URL is of the form 'NAME.git'
-    OR
-     type=github repo=opensciencegrid/cvmfs-config-osg tag=0.1 hash=e2b54cd1b94c9e3eaee079490c9d85f193c52249
-    'name' can be taken from the repo if not specified.
-
-    If nocheck is True, hashes do not have to match.
-    """
-    contents = {}
-    for entry in line.split():
-        info = entry.split("=", 1)
-        if len(info) > 1:
-            contents[info[0].strip()] = info[1].strip()
-
-    name = contents.get("name")
-    tag_type = contents.get("type", "")
-    if tag_type == "github":
-        repo = contents.get("repo")
-        if not repo:
-            raise Error("No repo specified: %s" % line)
-        m = re.match(r"([^\s/]+)/([^\s/]+?)(?:.git)?$", repo)
-        if not m:
-            raise Error("Repo syntax must be owner/project: %s" % line)
-        owner, project = m.group(1, 2)
-        git_url = "https://github.com/%s/%s" % (owner, project)
-        if not name:
-            name = project
-
-    elif tag_type == "git":
-        git_url = contents.get('url')
-        if not git_url:
-            raise Error("No git URL provided: %s" % line)
-        if not name:
-            basename = os.path.basename(git_url)
-            if basename[-4:] == '.git':
-                name = basename[:-4]
-            else:
-                raise Error("No package name specified: %s" % line)
+# common fetch options not found in .source line
+FetchOptions = collections.namedtuple('FetchOptions',
+    ['destdir', 'cache_prefix', 'nocheck', 'want_spec']
+)
 
-    else:
-        raise Error("Only 'git'- and 'github'-type sources are understood: %s" % line)
+# fetch handlers are defined like:
+#
+#   fetch_xyz_source(
+#       required_named_or_positional_arg, ...,
+#       optional_named_or_positional_arg=None, ...,
+#       ops=None,  # no positional args allowed after 'ops'; only named fields
+#       named_arg=None, ...,
+#       **kw  # only list if extra args are intended (to pass to another fn)
+#   )
 
-    git_hash = contents.get("hash")
-    if not git_hash:
-        raise Error("git hash not provided.")
+def fetch_cached_source(relpath, sha1sum=None, ops=None):
+    uri = os.path.join(ops.cache_prefix, relpath)
+    return fetch_uri_source(uri, sha1sum, ops=ops)
+
+def fetch_uri_source(uri, sha1sum=None, ops=None, filename=None):
+    if uri.startswith('/'):
+        uri = "file://" + uri
+        log.warning("Absolute path names in .source files break the 4th wall")
```

Comment:
I don't think 4th wall means what you think it means

Reply:
>I don't think 4th wall means what you think it means

Inconceivable!

Reply:
What i meant was, if you are looking at a .source file that contains an absolute path, it's as if the text in that source file waves to your face, and then reaches out of your monitor, grabs your wallet right out of your pocket, and removes the photo of your mom.

If that's not breaking the 4th wall, idk what is.

Reply:
OK seriously, could you change this warning into something I don't have to explain?